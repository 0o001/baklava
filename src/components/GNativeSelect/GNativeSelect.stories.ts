import { storiesOf } from '@storybook/vue';
import GNativeSelect from './GNativeSelect.vue';

storiesOf('GNativeSelect', module)
  .addParameters({ component: GNativeSelect })
  .add('Default', () => ({
    components: { GNativeSelect },
    props: {
      options: {
        default: [
          {
            value: null,
            text: 'Placeholder',
            disabled: true,
            hidden: true,
          },
          {
            value: 'Value1',
            text: 'Text1',
          },
          {
            value: 'Value2',
            text: 'Text2',
          },
        ],
      },
    },
    data () {
      return {
        selectedItem: {
          value: 'Value2',
          text: 'Text2',
        },
        selectedItemWithPlaceholder: {},
      };
    },
    template: `
    <div style="margin: 40px 200px 200px">
      <GNativeSelect class="g-mb-20" :options="options" v-model="selectedItemWithPlaceholder" label="Email"  />
      <GNativeSelect class="g-mb-20" :options="options" v-model="selectedItem" label="Email"  />
      <GNativeSelect class="g-mb-20" :options="options" v-model="selectedItem" isBorderless label="Email" />
      <GNativeSelect class="g-mb-20" :options="options" v-model="selectedItem" label="Email" feedback="feedback" isOutlineLabel/>
      <GNativeSelect class="g-mb-20" :options="options" v-model="selectedItem" />
      Selected Item: {{ selectedItem.value }}
     </div>`,
  }));
