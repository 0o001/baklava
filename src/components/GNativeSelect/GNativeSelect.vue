<template>
  <div :class="wrapperClass">
    <GFieldWrapper
      :is-value="!!value"
      :is-outline-label="isOutlineLabel"
      :feedback="feedback"
      :label="label"
      icon="chevron-down"
      :success="success"
      :error="error"
      :disable="disable"
    >
      <select
        :value="value[valueKey] || options[0][valueKey]"
        @input="onInput"
        v-on="listeners"
        v-bind="$attrs"
      >
        <option
          v-for="option in options"
          :value="option[valueKey]"
          :key="option[valueKey]"
          :disabled="option.disabled"
          :hidden="option.hidden"
        >
          {{ option[textKey] }}
        </option>
        <select />
      </select>
    </GFieldWrapper>
  </div>
</template>

<script lang="ts">

import ClickOutside from '../../directives/ClickOutside.ts';
import GFieldWrapper from '../GFieldWrapper/GFieldWrapper.vue';

export default {
  name: 'GNativeSelect',
  components: {
    GFieldWrapper,
  },
  directives: {
    ClickOutside,
  },
  props: {
    type: {
      default: 'text',
      type: String,
    },
    isOutlineLabel: {
      default: false,
      type: Boolean,
    },
    isBorderless: {
      default: false,
      type: Boolean,
    },
    success: {
      default: false,
      type: Boolean,
    },
    error: {
      default: false,
      type: Boolean,
    },
    disable: {
      default: false,
      type: Boolean,
    },
    label: {
      default: '',
      type: String,
    },
    feedback: {
      default: '',
      type: String,
    },
    valueKey: {
      type: String,
      default: 'value',
    },
    textKey: {
      type: String,
      default: 'text',
    },
    options: {
      required: true,
      type: Array,
    },
    value: {
      type: Object,
      default: () => ({}),
    },
  },
  computed: {
    wrapperClass () {
      return {
        'g-native-select': true,
        '-borderless': this.isBorderless,
      };
    },
    listeners () {
      const { input, ...rest } = this.$listeners;

      return rest;
    },
  },
  methods: {
    onInput (event: any) {
      const selectedItem = this.options.find((item: any) => (item[this.valueKey] || '').toString() === event.target.value);

      if (selectedItem) {
        this.$emit('input', selectedItem);
      }
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-native-select {
    ::v-deep .g-field-wrapper {
      .content select {
        -webkit-appearance: none;
        z-index: 2;
        color: var(--main-grey-500);
      }

      &.big .content {
        height: 38px;
      }

      &.small .content {
        height: 32px;
      }
    }

    &.-borderless {
      ::v-deep .g-field-wrapper {
        label {
          background: linear-gradient(180deg, white 50%, var(--bg-grey-500) 50%);
        }

        .content {
          border: none;
          background: var(--bg-grey-500);
          color: var(--main-grey-500);
        }
      }
    }
  }
</style>
