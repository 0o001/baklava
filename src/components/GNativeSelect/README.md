# GNativeSelect

## Props

| Name                 | Type      | Description | Default      |
| -------------------- | --------- | ----------- | ------------ |
| `type`               | `String`  |             | `"text"`     |
| `is-outline-label`   | `Boolean` |             | `false`      |
| `is-borderless`      | `Boolean` |             | `false`      |
| `success`            | `Boolean` |             | `false`      |
| `error`              | `Boolean` |             | `false`      |
| `disable`            | `Boolean` |             | `false`      |
| `label`              | `String`  |             | `""`         |
| `feedback`           | `String`  |             | `""`         |
| `value-key`          | `String`  |             | `"value"`    |
| `text-key`           | `String`  |             | `"text"`     |
| `options` *required* | `Array`   |             |              |
| `v-model`            | `Object`  |             | `() => ({})` |

## Computed Properties

| Name           | Type      | Description                      |
| -------------- | --------- | -------------------------------- |
| `wrapperClass` | `object`  | **Dependencies:** `isBorderless` |
| `listeners`    | `unknown` | **Dependencies:** `$listeners`   |

## Events

| Name    | Description                                                                                                                                             |
| ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `input` | <br/>**Arguments**<br/><ul><li>**`this.options.find((item: any) => (item[this.valueKey] || '').toString() === event.target.value): unknown`**</li></ul> |

## Methods

### onInput()

**Syntax**

```typescript
onInput(event: any): void
```

