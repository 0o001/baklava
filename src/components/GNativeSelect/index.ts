import GNativeSelect from './GNativeSelect.vue';

export default GNativeSelect;
