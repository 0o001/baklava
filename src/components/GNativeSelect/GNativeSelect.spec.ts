import { mount, Wrapper } from '@vue/test-utils';
import GNativeSelect from '.';

describe('GNativeSelect', () => {
  let wrapper: Wrapper<any>;

  it('should match snapshot when show is prop is given as false ', () => {
    wrapper = mount(GNativeSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render given defaultValue when variant prop set', () => {
    wrapper = mount(GNativeSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        defaultValue: 'Value2',
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({ '-borderless': false, 'g-native-select': true });
    expect(wrapper.find('select').element.getAttribute('defaultValue')).toEqual('Value2');
  });

  it('should render given isBorderless when variant prop set', () => {
    wrapper = mount(GNativeSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        isBorderless: true,
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({ '-borderless': true, 'g-native-select': true });
  });
});
