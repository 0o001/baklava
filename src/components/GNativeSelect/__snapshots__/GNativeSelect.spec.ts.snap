// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GNativeSelect should match snapshot when show is prop is given as false  1`] = `
<div
  class="g-native-select"
>
  <div>
    <div
      class="g-field-wrapper -borderline -top big"
    >
      <!---->
       
      <div
        class="content"
      >
        <select>
          <option
            value="Value1"
          >
            
        Text1
      
          </option>
          <option
            value="Value2"
          >
            
        Text2
      
          </option>
           
          <select />
        </select>
         
        <svg
          class="g-icon feather feather--chevron-down"
          dataName="chevron-down"
          dataTags=""
          dataType="chevron-down"
          fill="none"
          height="24px"
          stroke="var(--light-grey-500)"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
       
      <!---->
    </div>
  </div>
</div>
`;
